codetransformer/decompiler.py

```diff
@@ -0,0 +1,1402 @@
+class MakeFunctionContext(immutable):
+    __slots__ = ('closure',)
+
+
+def pycode_to_body(co, context):
+    """
+    Convert a Python code object to a list of AST body elements.
+    """
+    code = Code.from_pycode(co)
+    body = instrs_to_body(deque(code.instrs), context)
+    if context.in_function:
+        return make_global_and_nonlocal_decls(code.instrs) + body
+    return body
+
+
+def instrs_to_body(instrs, context):
+    """
+    Convert a list of Instruction objects to a list of AST body nodes.
+    """
+    stack = []
+    body = []
+    process_instrs(instrs, stack, body, context)
+
+    if stack:
+        raise DecompilationError(
+            "Non-empty stack at the end of instrs_to_body(): %s." % stack
+        )
+    return body
+
+
+def process_instrs(queue, stack, body, context):
+    """
+    Process instructions from the instruction queue.
+    """
+    next_instr = queue.popleft
+    while queue:
+        newcontext = _process_instr(next_instr(), queue, stack, body, context)
+        if newcontext is not None:
+            context = newcontext
+
+
+@singledispatch
+def _process_instr(instr, queue, stack, body, context):
+    raise AssertionError(
+        "process_instr() passed a non-instruction argument %s" % type(instr)
+    )
+
+
+@_process_instr.register(instrs.Instruction)
+def _instr(instr, queue, stack, body, context):
+    raise DecompilationError(
+        "Don't know how to decompile instructions of type %s" % type(instr)
+    )
+
+
+@_process_instr.register(instrs.EXTENDED_ARG)
+def _process_instr_extended_arg(instr, queue, stack, body, context):
+    """We account for EXTENDED_ARG when constructing Code objects."""
+    pass
+
+
+@_process_instr.register(instrs.UNPACK_SEQUENCE)
+def _process_instr_unpack_sequence(instr, queue, stack, body, context):
+    body.append(make_assignment(instr, queue, stack))
+
+
+@_process_instr.register(instrs.IMPORT_NAME)
+def _process_instr_import_name(instr, queue, stack, body, context):
+    """
+    Process an IMPORT_NAME instruction.
+
+    Side Effects
+    ------------
+    Pops two instuctions from `stack`
+    Consumes instructions from `queue` to the end of the import statement.
+    Appends an ast.Import or ast.ImportFrom node to `body`.
+    """
+    # If this is "import module", fromlist is None.
+    # If this this is "from module import a, b fromlist will be ('a', 'b').
+    fromlist = stack.pop().arg
+
+    # (optimization) level argument to __import__.  Should be 0, 1, or 2.
+    # Doesn't affect the AST.
+    level = stack.pop().arg
+
+    module = instr.arg
+    if fromlist is None:  # Regular import.
+        _pop_import_LOAD_ATTRs(module, queue)
+        store = queue.popleft()
+        asname = store.arg
+        body.append(
+            ast.Import(
+                names=[
+                    ast.alias(
+                        name=module,
+                        asname=(
+                            asname if asname != module.split('.')[0] else None
```

Comment:
The reason this check is here, btw, is because `import a.b.c` generates a `STORE_NAME` to `a`, but doesn't have an alias node in the ast.  So what we actually need to do is check if we popped any LOAD_ATTRs.